osutil/epoll/epoll.go

```diff
@@ -0,0 +1,177 @@
+	registeredFdCount int
+	countMutex        sync.Mutex
+}
+
+// Open opens an event monitoring descriptor.
+func Open() (*Epoll, error) {
+	fd, err := unix.EpollCreate1(unix.EPOLL_CLOEXEC)
+	if err != nil {
+		return nil, fmt.Errorf("cannot open epoll file descriptor: %w", err)
+	}
+	e := &Epoll{
+		fd:                fd,
+		registeredFdCount: 0,
+	}
+	runtime.SetFinalizer(e, func(e *Epoll) {
+		if e.fd != -1 {
+			e.Close()
+		}
+	})
+	return e, nil
+}
+
+// Close closes the event monitoring descriptor.
+func (e *Epoll) Close() error {
+	runtime.SetFinalizer(e, nil)
+	e.countMutex.Lock()
+	fd := e.fd
+	e.fd = -1
+	e.registeredFdCount = 0
+	e.countMutex.Unlock()
+	return unix.Close(fd)
+}
+
+// Fd returns the integer unix file descriptor referencing the open file.
+func (e *Epoll) Fd() int {
+	return e.fd
+}
+
+// RegisteredFdCount returns the number of file descriptors which are currently
+// registered to the epoll instance.
+func (e *Epoll) RegisteredFdCount() int {
+	return e.registeredFdCount
+}
+
+// Register registers a file descriptor and allows observing speicifc I/O readiness events.
+//
+// Please refer to epoll_ctl(2) and EPOLL_CTL_ADD for details.
+func (e *Epoll) Register(fd int, mask Readiness) error {
+	err := unix.EpollCtl(e.fd, unix.EPOLL_CTL_ADD, fd, &unix.EpollEvent{
+		Events: uint32(mask),
+		Fd:     int32(fd),
+	})
+	if err != nil {
+		return err
+	}
+	e.countMutex.Lock()
+	e.registeredFdCount += 1
+	e.countMutex.Unlock()
+	runtime.KeepAlive(e)
+	return err
+}
+
+// Deregister removes the given file descriptor from the epoll instance.
+//
+// Please refer to epoll_ctl(2) and EPOLL_CTL_DEL for details.
+func (e *Epoll) Deregister(fd int) error {
+	err := unix.EpollCtl(e.fd, unix.EPOLL_CTL_DEL, fd, &unix.EpollEvent{})
+	if err != nil {
+		return err
+	}
+	e.countMutex.Lock()
+	e.registeredFdCount -= 1
+	e.countMutex.Unlock()
+	runtime.KeepAlive(e)
+	return err
+}
+
+// Modify changes the set of monitored I/O readiness events of a previously registered file descriptor.
+//
+// Please refer to epoll_ctl(2) and EPOLL_CTL_MOD for details.
+func (e *Epoll) Modify(fd int, mask Readiness) error {
+	err := unix.EpollCtl(e.fd, unix.EPOLL_CTL_MOD, fd, &unix.EpollEvent{
+		Events: uint32(mask),
+		Fd:     int32(fd),
+	})
+	runtime.KeepAlive(e)
+	return err
+}
+
+// Event describes an IO readiness event on a specific file descriptor.
+type Event struct {
+	Fd        int
+	Readiness Readiness
+}
+
+// WaitTimeout blocks and waits with the given timeout for arrival of events on any of the added file descriptors.
+//
+// A msec value of -1 disables timeout.
+//
+// Please refer to epoll_wait(2) and EPOLL_WAIT for details.
+//
+// Warning, using epoll from Golang explicitly is tricky.
+func (e *Epoll) WaitTimeout(duration time.Duration) ([]Event, error) {
+	msec := int(duration.Milliseconds())
+	if duration < 0 {
+		msec = -1
+	}
+	sysEvents := make([]unix.EpollEvent, e.registeredFdCount)
+	n, err := unix.EpollWait(e.fd, sysEvents, msec)
```

Comment:
I looked a bit more (also checked netpoll_epoll.go from the go runtime for reference) how to deal with EINTR - I think the easiest way out is to simply kill WaitTimeout() until we need it (AIUI the prompting code is now just using Wait() ?).

This means we can cheat and deal with EINTR simply by ignoring it :) I.e.
```go
        var n int
	for {
		n, err = unix.EpollWait(e.fd, sysEvents, -1)
		if err != nil {
			if err == syscall.EINTR {
				continue
			}
			return nil, err
		}
	}
```
if we keep the timeout we will have to keep track of how much time already elapsed and restart the syscall with the new timeout. Which is maybe fine too, I let you play with it :)

Reply:
Thanks so much for looking into this! I was thinking the same thing (loop until !EINTR). I'll try adding it and tracking elapsed time to make EINTR disappear from the caller point of view.

Reply:
Done! Local testing suggests the fix should work. I tried to make it as concise as possible, hopefully it looks good to you.